### ISSUE TITLE
Lift unnecessary restriction on CAS failure ordering

### ISSUE THREAD

**Issue by tmiasko** [rust-lang/rust#68464]
Currently `compare_exchange` requires the failure ordering to "be equivalent to
or weaker than a success ordering". On the other hand C11/C++11 requires only
that "failure shall be no stronger than the success", which arguably means that
one can write e.g. `compare_exchange(..., Release, Acquire)` in C/C++ but not in Rust.

Arguably, because neither C11 standard nor C++11 standard defines what it means
for an ordering to be stronger from another. When the issue was raised in
[LWG2445], the proposed and accepted resolution was to lift those restrictions
altogether, leaving only requirement that "the failure argument shall not be
`memory_order_release` nor `memory_order_acq_rel`". 

It would be beneficial to remove success/failure ordering restrictions for
reasons described in C++ proposal [P0418r2].

[LWG2445]: https://cplusplus.github.io/LWG/issue2445
[P0418r2]: http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2016/p0418r2.html

EDIT: Restrictions were lifted in clang & LLVM:

* https://reviews.llvm.org/rG431e3138a1f3fd2bb7b25e1f4766d935058136f8
* https://reviews.llvm.org/rG819e0d105e84c6081cfcfa0e38fd257b6124553a

**Comment by tmiasko** [comment]
Blocked on https://bugs.llvm.org/show_bug.cgi?id=33332. The cmpxchg release acquire is accepted by LLVM but miscompiled on AArch64.

**Comment by m-ou-se** [comment]
@tmiasko I'm about to send a PR that implements this.

**Comment by tmiasko** [comment]
@m-ou-se I do have implementation ready as well, but that's fine, I can review instead :-).

**Comment by m-ou-se** [comment]
I split it in a PR that renames the intrinsics first, and a separate PR for exposing the new ordering combinations afterwards.

**Comment by m-ou-se** [comment]
https://github.com/rust-lang/rust/pull/98385

**Comment by m-ou-se** [comment]
To do:

- [ ] Publicly expose the new orderings.
- [ ] Add codegen test.
- [ ] Update memory order lint to accept the new orderings.
